kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/PulsarStorageProducerIdManagerImpl.java

```diff
@@ -0,0 +1,210 @@
+        // please note that the read operation is async,
+        // and it is not execute inside this synchronized block
+        CompletableFuture<Reader<byte[]>> readerHandle = ensureReaderHandle();
+        final CompletableFuture<Void> newReadHandle =
+                readerHandle.thenCompose(this::readNextMessageIfAvailable);
+        currentReadHandle = newReadHandle;
+        return newReadHandle.whenComplete((a, b) -> {
+            endReadLoop(newReadHandle);
+            if (b != null) {
+                throw new CompletionException(b);
+            }
+        });
+    }
+
+    private synchronized void endReadLoop(CompletableFuture<?> handle) {
+        if (handle == currentReadHandle) {
+            currentReadHandle = null;
+        }
+    }
+
+    @Override
+    public synchronized CompletableFuture<Long> generateProducerId() {
+        Long booked = availableIdsLocally.poll();
+        if (booked != null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Returning pre-allocated id {} for {}", booked, topic);
+            }
+            return CompletableFuture.completedFuture(booked);
+        }
+        if (log.isDebugEnabled()) {
+            log.debug("Allocating new block of ids for {}", topic);
+        }
+        CompletableFuture<Producer<byte[]>> producerHandle = pulsarClient.newProducer()
+                .enableBatching(false)
+                .topic(topic)
+                .accessMode(ProducerAccessMode.WaitForExclusive)
+                .blockIfQueueFull(true)
+                .createAsync();
+        return producerHandle.thenCompose(opProducer -> {
+            // nobody can write now to the topic
+            // wait for local cache to be up-to-date
+            CompletableFuture<Long> dummy = ensureLatestData(true)
+                    .thenCompose((___) -> {
+                        long start = nextId.get();
+                        List<Long> block = generateBlock(start + blockSize, blockSize);
+                        final long nextAvailableId = start + blockSize;
+                        // write to Pulsar
+                        byte[] serialized = BigInteger.valueOf(nextAvailableId).toByteArray();
+                        CompletableFuture<Long> res = opProducer
+                                .newMessage()
+                                .key("") // always the same key, this way we can rely on compaction
+                                .value(serialized)
+                                .sendAsync()
+                                .thenApply((msgId) -> {
+                                    if (log.isDebugEnabled()) {
+                                        log.debug("{} written {} as {}", this, nextAvailableId, msgId);
+                                    }
+                                    nextId.set(nextAvailableId);
+                                    availableIdsLocally.addAll(block);
+                                    long result = availableIdsLocally.remove();
```

Comment:
Should these two lines be composed into an atomic operation? I see `availableIdsLocally.poll()` could be called in another thread?

Reply:
in theory there can be only 1 Exclusive producer working on the topic (partition) at a time. Even in case of multiple executions in the same JVM.
So there should be no other thread accessing `availableIdsLocally` when we are here